### PR TITLE
[MODTEMGENG-14] Forbid in use templates deletion

### PR DESCRIPTION
**Pull request by opochepam**
## Purpose
Resolves: MODTEMPENG-14
Not allow user to delete patron notice template thru API (backend) if currently in use
## Links
https://issues.folio.org/browse/MODTEMPENG-14

### CODE REVIEW

File: src/main/java/org/folio/template/service/TemplateServiceImpl.java [hunk 1]
```diff
@@ -67,14 +77,21 @@ public TemplateServiceImpl(Vertx vertx, OkapiConnectionParams params) {
...
   }
 
   @Override
-  public Future<Boolean> deleteTemplate(String id) {
-    return templateDao.deleteTemplate(id);
+  public Future<Boolean> deleteTemplate(String id, OkapiConnectionParams params) {
+
+    String query = format("loanNotices=/@templateId %1$s " +
+      "OR requestNotices=/@templateId %1$s " +
+      "OR requestNotices=/@templateId %1$s", id);
```
[the remark this answers is not in the document]
Comment (opochepam) [reply src/main/java/org/folio/template/service/TemplateServiceImpl.java]: Yes, you're right, thanks.

File: src/test/java/org/folio/rest/impl/TemplateDeletionTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  private static void setupStub() {
+
+    String query = "loanNotices=/@templateId %1$s " +
+      "OR requestNotices=/@templateId %1$s " +
+      "OR requestNotices=/@templateId %1$s";
+
+    wireMockServer.stubFor(
+      get(urlPathEqualTo("/patron-notice-policy-storage/patron-notice-policies"))
+        .withQueryParam("query", equalTo(format(query, inUseTemplate.getString("id"))))
+        .withQueryParam("limit", equalTo("0"))
```
[the remark this answers is not in the document]
Comment (opochepam) [reply src/test/java/org/folio/rest/impl/TemplateDeletionTest.java]: limit 0 will return empty array, but the limit doest not affect `totalRecords` which I use to check if there are any records matching the query

[the remark this answers is not in the document]
Comment (dmytrotur) [reply src/test/java/org/folio/rest/impl/TemplateDeletionTest.java]: Makes sense, thanks for the explanation.

File: src/main/java/org/folio/template/client/CirculationStorageClient.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    webClient.getAbs(params.getOkapiUrl() + "/patron-notice-policy-storage/patron-notice-policies")
+      .addQueryParam("query", query)
+      .addQueryParam("limit", String.valueOf(limit))
+      .putHeader(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON)
+      .putHeader(RestVerticle.OKAPI_HEADER_TENANT, params.getTenant())
+      .putHeader(RestVerticle.OKAPI_HEADER_TOKEN, params.getToken())
+      .send(future.completer());
+
+    return future.map(resp -> resp.body().toJsonObject());
```
[the remark this answers is not in the document]
Comment (opochepam) [reply src/main/java/org/folio/template/client/CirculationStorageClient.java]: added the check

File: src/test/java/org/folio/rest/impl/TemplateDeletionTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      .delete("/templates/" + inUseTemplate.getString("id"))
+      .then()
+      .statusCode(400);
+  }
+
+  private static void setupStub() {
+
+    String query = "loanNotices=/@templateId %1$s " +
+      "OR requestNotices=/@templateId %1$s " +
+      "OR requestNotices=/@templateId %1$s";
```
[the remark this answers is not in the document]
Comment (opochepam) [reply src/test/java/org/folio/rest/impl/TemplateDeletionTest.java]: fixed

File: src/main/java/org/folio/template/util/TemplateEngineHelper.java [hunk 5]
```diff
@@ -27,12 +28,21 @@ public static Response mapExceptionToResponse(Throwable throwable) {
...
       return Response.status(HttpStatus.SC_NOT_FOUND)
         .type(MediaType.TEXT_PLAIN)
         .entity(throwable.getMessage())
         .build();
     }
+
+    if (throwable instanceof InUseTemplateException) {
+      return Response.status(400)
+        .type(MediaType.TEXT_PLAIN_TYPE)
+        .entity("Cannot delete the template which is currently in use")
```
Comment (kostyantyn-kh) [inline src/main/java/org/folio/template/util/TemplateEngineHelper.java]:
```suggestion
        .entity("Cannot delete template which is currently in use")
```

File: descriptors/ModuleDescriptor-template.json [hunk 6]
```diff
@@ -35,7 +40,9 @@
         {
           "methods": ["DELETE"],
           "pathPattern": "/templates/{templateId}",
-          "permissionsRequired": ["templates.item.delete"]
+          "permissionsRequired": ["templates.item.delete"],
+          "modulePermissions": ["circulation-storage.patron-notice-policies.collection.get"]
```
Comment (kostyantyn-kh) [inline descriptors/ModuleDescriptor-template.json]: Minor version update is required due to permission changes

**Pull request merged**